### PR TITLE
[TACHYON-295] Only attempt to read from worker memory once per block stream.

### PR DESCRIPTION
**Pull request by calvinjia**
Since multiple reads will most likely occur for each block, trying to find a copy of the block in a worker's memory is very expensive each time and unlikely to succeed after the first failure.

The performance improvement is around 240s -> 10s when reading a 1.5 gb file in HDFS on my local machine.

### CODE REVIEW

**Comment by haoyuan** [comment]
@shimingfei @carsonwang @GraceH

File: core/src/main/java/tachyon/client/RemoteBlockInStream.java [hunk 1]
```diff
@@ -206,6 +212,10 @@ public int read(byte[] b, int off, int len) throws IOException {
       bytesLeft -= bytesToRead;
       mBlockPos += bytesToRead;
     }
+
+    // Unable to read from worker memory, reading this block from underfs.
+    mAttemptReadFromWorkers = false;
```
Comment (carsonwang) [inline core/src/main/java/tachyon/client/RemoteBlockInStream.java]: Should this line be placed in the if (bytesLeft > 0) ?

**Comment by haoyuan** [comment]
Jenkins, test this please.

File: core/src/main/java/tachyon/client/RemoteBlockInStream.java (hunk 1 above)
Comment (calvinjia) [inline core/src/main/java/tachyon/client/RemoteBlockInStream.java]: Good catch, thanks.

**Comment by shimingfei** [comment]
@calvinjia In the testing, what is the block size and how many times is the file read ?
it seems that, only a file which lays on under file system is read multi - times with the same FileInStream instance will benefit from this change.

**Comment by calvinjia** [comment]
@shimingfei Yeah you are right. The block size here is the Tachyon block size or underfs block size, so it will be relatively large, ie 64M - 1GB. For a spark job, the client will request small parts of the block each time, ie 64k. One way to make the issue less severe is to increase the read buffer size, but that is just a work around and we shouldn't expect that from all clients.

**Comment by shimingfei** [comment]
Ok, I see, Thanks!
+1 LGTM

**Comment by carsonwang** [comment]
+1 LGTM
One thing not relevent to this PR I found while doing test is that, reading from a remote worker's memory could be much slower than reading from UFS directly. This is ture even if the client is not on a HDFS datanode.

**Comment by haoyuan** [comment]
@carsonwang  Any number for that case?

**Comment by carsonwang** [comment]
In my test, it takes 16-17s for reading 1GB from a remote worker's memory. Most of the overhead is caused by network. If the tachyon client is on a datanode where all blocks of the HDFS file in on the node, it only takes 3.2 seconds reading 1GB data from HDFS. One a namenode, it takes about 10s reading from HDFS which is still faster than reading from a remote worker's memory.

**Comment by haoyuan** [comment]
What's the network environment?

**Comment by haoyuan** [comment]
1Gb ethernet?

**Comment by carsonwang** [comment]
Yes, it is 1GB/sec.

**Comment by haoyuan** [comment]
I suppose you meant 1Gb/sec instead of 1GB/sec. In that case, transferring 1GB data would take at least 8 seconds. If the network is 10Gb/sec, the transfer time would be reduced by 8X theoretically. It could be a good optimization to choose where to load data if it's not in local memory.

+ @ooq

**Comment by carsonwang** [comment]
Sorry for the typo. Yes, it is 1Gb/sec. This sounds reasonable as it takes 10 seconds for me reading 1GB data from all the datanodes. I am checking if any netty configurations that may affect the speed when reading from a remote worker's memory.

**Comment by carsonwang** [comment]
@haoyuan I figured this out. The configuration "tachyon.user.remote.read.buffer.size.byte" has a big effect on the performance. The default setting is 1MB. For reading 1GB data from a remote worker, the client need send 1000+ requests. So it takes about 17 seconds. If I set this buffer size higher (i.e. 8MB), I can read 1GB data from a remote worker more quickly (about 10.5 seconds). Is the default value 1MB appropriate for most of the other cases?

**Comment by haoyuan** [comment]
@carsonwang  This is great. Please submit a PR to 0.6 branch (let's use 8MB for now since it is mainly large sequential read currently). I will merge that to the master branch.

**Pull request merged**